### PR TITLE
Add API documentation

### PR DESCRIPTION
**Pull request by vasilisp**
`ReactiveData` gets API documentation. Let me know if there is anything unclear, or that needs expansion.

### CODE REVIEW

File: src/reactiveData.mli [hunk 1]
```diff
@@ -17,66 +17,213 @@
...
+      updated whenever [a] and [b] change *)
   val concat : 'a t -> 'a t -> 'a t
+
+  (** [rev a] is the reversal of [a]; [rev a] gets updated along with
+      [a] *)
   val rev : 'a t -> 'a t
+
+  (**/**)
   val sort : ('a -> 'a -> int) -> 'a t -> [`Not_implemented]
   val filter : ('a -> unit) -> 'a t -> [`Not_implemented]
```
Comment (Drup) [inline src/reactiveData.mli]: Please just remove that.

File: src/reactiveData.mli [hunk 2]
```diff
@@ -17,66 +17,216 @@
...
+      complexity. *)
   type 'a p =
-      I of int * 'a
-    | R of int
-    | U of int * 'a
-    | X of int * int
+    | I of int * 'a  (** [I i v] adds [v] at position [i] *)
+    | R of int       (** [R i] removes [i]-th element *)
+    | U of int * 'a  (** [U i v] substitutes [i]-th element with [v] *)
+    | X of int * int (** [X i j] swaps the [i]-th and [j]-th elements *)
```
Comment (Drup) [inline src/reactiveData.mli]: `I (i,v)` and all the others.

File: src/reactiveData.mli [hunk 3]
```diff
@@ -17,66 +17,216 @@
...
+      length and n is the current size of the list. (Arbitrary patches
+      are slower, requiring O(m * n).) *)
+  type 'a patch = 'a p list
+
   include S with type 'a data = 'a list
-             and type 'a patch = 'a p list
+             and type 'a patch := 'a patch
 
+  (** Deprecated. Use [empty]. *)
   val nil : 'a t
```
Comment (Drup) [inline src/reactiveData.mli]: Can we just remove it ?

File: src/reactiveData.mli [hunk 4]
```diff
@@ -17,66 +17,216 @@
...
+
+  (** Add element to the end *)
   val append : 'a -> 'a handle -> unit
```
Comment (Drup) [inline src/reactiveData.mli]: That's the wrong name, it should be called `snoc`.

File: src/reactiveData.mli [hunk 5]
```diff
@@ -17,66 +17,216 @@
...
+
+      The result is a signal of value [f m_n (f ... (f m_1 v))], where
+      [m_1] ... [m_n] are the messages that have been applied since
+      the beginning of [fold]. [m_1] is a pseudo-message [Set l],
+      accounting for the contents [l] of [c] at the time when
+      accumulation starts. *)
   val fold : ('a -> 'b msg -> 'a) -> 'b t -> 'a -> 'a React.signal
+
+  (** Signal corresponding to contents *)
   val value_s : 'a t -> 'a data React.S.t
```
Comment (Drup) [inline src/reactiveData.mli]: By symmetry with `event`, I think this should be called `signal`.

File: src/reactiveData.mli [hunk 6]
```diff
@@ -17,66 +17,216 @@
...
+  (** Handle that permits applying incremental updates *)
   type 'a handle
-  type 'a t
+
+  (** Empty data structure *)
   val empty : 'a t
+
+  (** Build a container from initial contents. The handle can be used
+      for performing reactive updates. *)
   val make : 'a data -> 'a t * 'a handle
```
Comment (Drup) [inline src/reactiveData.mli]: By analogy with `React.S.create`, this should be called `create`.

File: src/reactiveData.mli [hunk 7]
```diff
@@ -17,66 +17,216 @@
...
   val make_from : 'a data -> 'a msg React.E.t -> 'a t
+
+  (** Convert a React signal into a ReactiveData container.
+
+      Whenever the signal changes from value [v] to value [v'], we
+      detect the differences between [v] and [v'], and perform
+      downstream computation (e.g., for [map]) only on the new and
+      modified elements. *)
   val make_from_s :
     ?eq:('a -> 'a -> bool) -> 'a data React.S.t -> 'a t
```
Comment (Drup) [inline src/reactiveData.mli]: `from_signal` ?

File: src/reactiveData.mli [hunk 8]
```diff
@@ -17,66 +17,216 @@
...
+  (** Empty data structure *)
   val empty : 'a t
+
+  (** Build a container from initial contents. The handle can be used
+      for performing reactive updates. *)
   val make : 'a data -> 'a t * 'a handle
+
+  (** [make_from d e] is a container whose initial value is [d], and
+      which gets updated for every occurrence of [e] *)
   val make_from : 'a data -> 'a msg React.E.t -> 'a t
```
Comment (Drup) [inline src/reactiveData.mli]: `from_event` ? This way  there is a correspondance between event/signal and from_event/from_signal.

**Comment by Drup** [comment]
The documentation is good. I just proposed a set of renaming of the functions, because those names are just terrible and inconsistent.

File: src/reactiveData.mli (hunk 3 above)
Comment (vasilisp) [inline src/reactiveData.mli]: I just removed the only use of `RList.nil` known to mankind, so yes we can.

**Comment by vasilisp** [comment]
I am not a fan of the old names, I like your suggestions, and now is the right time to change them. We can easily update Eliom + JSOO, and hopefully our BeSport friends won't terribly mind the breakage. Is there any other user?

**Comment by Drup** [comment]
Yes, there are other users (@hnrgrgr, @smondet, that I know of, probably some more). But in this case, I really want to change them. I want a clean base for us to add things later.

File: src/reactiveData.mli [hunk 9]
```diff
@@ -17,66 +17,216 @@
...
+  (** Produce container list containing a single, constant element *)
   val singleton : 'a -> 'a t
+
+  (** Produce reactive list containing a single element that gets
+      updated based on a signal *)
   val singleton_s : 'a React.S.t -> 'a t
+
+  (** [concat a b] is the concatenation of [a] and [b], and it gets
+      updated whenever [a] and [b] change *)
   val concat : 'a t -> 'a t -> 'a t
```
Comment (Drup) [inline src/reactiveData.mli]: Ideally, this should be called `append`, but this one is a bit trickier (because there was an `append` before).

**Comment by hhugo** [comment]
Should I wait before make a Jsoo release ?

**Comment by vasilisp** [comment]
Yes, please wait. A patch for `Tyxml_js` is coming really soon.

**Comment by vasilisp** [comment]
The renaming happened in the `names` branch.

**Comment by vasilisp** [comment]
@hhugo @Drup, actually JSOO doesn't use any of the renamed functions. Unless we change something else, JSOO is OK.

**Comment by vasilisp** [comment]
Merged the documentation, not the name changes. Opening a new PR for these to make sure any concerns are expressed.

**Pull request merged**